### ISSUE TITLE
Add pagination to selection field types to avoid URI too long

### ISSUE THREAD

**Issue by mamazu** [sulu/sulu#6646]
| Q | A
| --- | ---
| Sulu Version | 2.3.7
| PHP Version | 7.4.12
| DB Version | mariadb:10.5.9
| Browser Version | Firefox 100.0.2

# Actual Behavior
When you want to create a list of Sulu pages, you use the `page_selection` property in Sulu. This gives you a User interface for selecting pages one by one and adding them. This way you can create a list of arbitrary length since the individual items are loaded once selected. If you save the page and come back to it Sulu tries to load all of the pages at once with a url like this:
```
http://localhost/admin/api/pages?locale=de_de&ids=b3750d14-5353-4170-88d9-e3073825f014,48461e37-dd7b-49b6-bed1-2624cb59e1dd,1667c2f1-c4fd-49e7-849e-10c250465705,e4c3e8fc-63aa-439f-aebf-a05175f1d6af,b410d373-a791-4bce-9f3f-c153672a8581,05babbb8-7a28-4c82-abb9-065f3a6658db,e2238c31-ebf7-4e79-a077-20ed551418f7,eb544ef8-1463-4dc6-8caa-405f98f7bb18,4c12e6a9-12a2-4dc5-8193-0f72e110fa3f,e751e1e0-0e00-41f6-9846-b021cb913174,eb6c9737-d3d2-41e6-8ab3-d5bb6885b99b,272e97d8-ac1f-47e6-a588-d60d0a2699c6,57e4f077-6a28-4ee4-b1c3-2f284040d51e
```

Generating a URL in that way will have the limitation of the maximum URL length. Doing some quick calculation:
```
Max URL length: 2048
Length of Id: 36 + 1 (for the comma as the seperator)
Prefix (/admin/api): 50

= (2048 - 50) / 37 = 54
```

Which means that the list is limited to 54 entries at most. Depending on the prefix. It's going to be enough in most cases. But a pagination would probably be better.

# Expected Behavior

**Either** when creating the list, stop the user from adding too many entries to it.
**Or** paginate the endpoint so that we don't need to load all things at first.

# Steps to Reproduce

* Create a page in your XML structure.
* Add the `page_selection` to it.
* Fill the page selection with more than 60 items (just to be sure)
* Save the page
* Go back to the main admin page
* Open the page you've just configured up again
* See the spinner load and in the console you get a `414 - URI too long`

# Possible Solutions

I don't know if it is easy to implement pagination but this would probably be the thing that scales best.

**Comment by alexander-schranz** [comment]
Thx for reporting.

A pagination of the selected Items I think would be great, think it is enough to just show the first 10 pages and then show the rest via a paginiation. But not sure how this will effect other cases as example the query for the pages. We would need a design to target this as we do not have any paginiation inside the selections / items list yet.

Still in most cases selection are only used for select some items and such a big selection is and edge case. For most other usecases I recommend using smart content based on tag or category filters, as that is mostly better manageable.

**Comment by alexander-schranz** [comment]
I tested it on our servers the 414 does not appear there so I think it has todo with limitation defined on your webserver.

I know that we normally increase already `large_client_header_buffers` in our webservers because it is already required for CMPs as they already write bigger cookie header as it is the default setting for the nginx. But also on our default server without large header buffers I did not run into 414 for this urls. What type of webserver do you use?

**Comment by mamazu** [comment]
I completely agree that this isn't the right or the normal way to use Sulu in this case. And the issues we are having also didn't occur at 56 Entries in the list. The page in question has more than 200 Entries in a list, so that will give the web server something to think about.

**Comment by mamazu** [comment]
Maybe something like this as a design suggestion:

![Untitled](https://user-images.githubusercontent.com/14860264/179834230-44ac390d-b1fe-46c2-99da-f010107e1a0f.png)

In the XML you could configure it like this with 10 being the default
```xml
<property name="pages" type="page_selection">
    <meta>
        <title lang="en">Pages</title>
        <title lang="de">Seiten</title>
    </meta>
    <params>
        <param name="page_size" value="10" />
    </params>
</property>
```

**Comment by mamazu** [comment]
I have tried to implement a UI for that but the problem is that if I want to change it in the `MultipleItemSelection`, I also have to change the logic how items are loaded. Because currently it shows only two items but in the background still loads all of them. Where would that code be located.

**Comment by alexander-schranz** [comment]
I'm not sure where this code currently lies would need to debug the selection. Know in the overlay there are LoadingStrategy used not and think the List should maybe be extended work in combination with `ids` and request only a limited numbers of them.

**Comment by mamazu** [comment]
I don't really have any good ideas on how to solve this. Every approach I've tried has some major downsides.

## Proper pagination
Adding a proper pagination to the list (as I intended to) needs changes in a lot of components (`MultipleItemSelection`, `Header`, newly created `Pagination`). Furthermore to properly implement this a lot of components would need to handle it as well. The downside I see with this version is changing the page with the pagination selector Sulu offers me to save the page. So I assume rendering a sublist would only save those items and all other pages of content will be lost. :frowning_face: 

## Paginated Requests
Only paginating the requests would be the only other option then. The problem here is that we can't do that in a general manner in the `ResourceRequester.getList` because we don't know what properties should be paginated. In the `TeaserSelection` it would be the `id` and in other components it might be another property. Another issue is that we also need dynamic paging sizes. Because if you have a component with UUIDs it can have much less UUIDs per request than it would be with normal ids.

## A possible solution
This brings me to the conclusion that the smartest way would be to change the `TeaserStore` to do the pagination. The problem with this is, that this solution isn't very generic and would need to be applied to all stores that should have "pagination".

I don't know. None of the solutions really sound like something satisfying. Please let me know if I am missing anything.
